### DISCUSSION TITLE
Provide the queryKey in error message

### DISCUSSION THREAD

**Discussion by arthurdenner** [TanStack/query#4042] [Ideas]
Hi! 👋🏽 

As one of the breaking changes on v4, the `queryKey` needs to be an Array. I want to report to our internal services when a non-array `queryKey` is provided so our microfrontends can migrate accordingly - we have 30+ of projects to update.

There's an error logged when a non-array queryKey is provided ([reference](https://github.com/TanStack/query/blob/f6eeab079d88df811fd827767db1457083e85b01/packages/query-core/src/query.ts#L356)) but it doesn't contain the provided queryKey. Is it possible to add it to the message?

My current solution is to override the `useQuery` export to check the `queryKey`, but I think doing it through the logger would be a cleaner solution.

Related question: `useQuery` doesn't throw, only logs, when a non-array `queryKey` is provided and it works fine, data is loaded, but other APIs such as `getQueryData` fail when receiving a string as `queryKey`, e.g.:

`index.js:1341 Uncaught TypeError: Cannot create property 'exact' on string 'someQueryKey'` ([reference](https://github.com/TanStack/query/blob/454c58481f6f5d5c3e182e6e418d65fdb4a9ec1f/packages/query-core/src/utils.ts#L153)) - `arg1` is the string, not falsy, therefore returned, which fails [here](https://github.com/TanStack/query/blob/454c58481f6f5d5c3e182e6e418d65fdb4a9ec1f/packages/query-core/src/queryCache.ts#L166).

Are there plans to throw an error in `useQuery` OR support strings in other APIs?

**Comment by TkDodo** [comment]
> Is it possible to add it to the message?

I've tried, but the problem is the overloads of useQuery and how we parse arguments:

https://github.com/tannerlinsley/react-query/blob/f9ebf9aaae59050331c22f143d835e46fac402cc/packages/query-core/src/utils.ts#L104-L121

So if it's neither an array, nor a function, it will be spread, and spreading a string results in the queryKey being `undefined` by the time we access it.

However, now that I'm thinking about it, maybe you want to contribute to add an additional dev-only runtime check similar to this one:

https://github.com/TanStack/query/blob/357ec041a6fcc4a550f3df02c12ecc7bcdefbc05/packages/query-core/src/query.ts#L353-L359

to `parseQueryArgs` and `parseMutationArgs`, that checks that if we get a string in and then logs a similar error? There, we would have access to the string then and could provide better messages.

> Are there plans to throw an error in useQuery OR support strings in other APIs?

if we throw in useQuery, it will go to error state and retry. I wanted to avoid that because it hides the issue. The uncaught type error seems to be a natural follow-up bug because we have no overloads for this function, thus we don't use `parseQueryArgs` to normalize. It's not an intentional throwing on our side.

**Reply by arthurdenner** [reply]
Thanks for the detailed reply, @TkDodo, especially on why not to throw in useQuery.
Rethinking what I suggested before, as it was a breaking change, I think it's fine how it is today.

I decided to [patch-package](https://www.npmjs.com/package/patch-package) v3 and report non-array keys in the `ensureQueryKeyArray` function ([reference](https://github.com/TanStack/query/blob/v3.39.2/src/core/utils.ts#L94)).
Once we don't have any reports, we know we can update safely.

Unless you still prefer to add more logs, I'm fine with closing the discussion.

**Comment by Djarma12** [comment]
I was having the same error, and problem was TYPO, where i write gueryKey instead a queryKey.
